ambry-commons/src/main/java/com.github.ambry.commons/NettyMetrics.java

```diff
@@ -0,0 +1,249 @@
+        (Gauge<Long>) () -> getNumHeapTotalActiveAllocations());
+    thisRegistry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectTotalAlloocations"),
+        (Gauge<Long>) () -> getNumDirectTotalAllocations());
+    thisRegistry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectTotalDealloocations"),
+        (Gauge<Long>) () -> getNumDirectTotalDeallocations());
+    thisRegistry.register(MetricRegistry.name(NettyMetrics.class, "NumberDirectTotalActiveAlloocations"),
+        (Gauge<Long>) () -> getNumDirectTotalActiveAllocations());
+  }
+
+  /**
+   * Start to collect metrics.
+   */
+  public void start() {
+    if (started.compareAndSet(false, true)) {
+      initializeConfig();
+      register();
+      collectorThread = Utils.newThread("netty-metrics", new NettyMetricCollector(), false);
+      collectorThread.start();
+      logger.trace("Thread is started to collect netty metrics");
+    }
+  }
+
+  /**
+   * Stop collecting metrics. This method would stop this singleton effectively. Do not reuse this single after
+   * calling this method.
+   */
+  public void stop() {
+    if (started.compareAndSet(true, false)) {
+      collectorThread.interrupt();
+      try {
+        stopLatch.await(config.nettyMetricsStopWaitTimeoutSeconds, TimeUnit.SECONDS);
+        logger.trace("Thread is stopped for collecting netty metrics");
+      } catch (InterruptedException e) {
+        logger.error("Failed to stop collector thread", e);
+      }
+    }
+  }
+
+  private class NettyMetricCollector implements Runnable {
+
+    @Override
+    public void run() {
+      while (true) {
+        PooledByteBufAllocatorMetric metric = PooledByteBufAllocator.DEFAULT.metric();
+        numDirectArenas = metric.numDirectArenas();
+        numHeapArenas = metric.numHeapArenas();
+        numThreadLocalCaches = metric.numThreadLocalCaches();
+        usedHeapMemory = metric.usedHeapMemory();
+        usedDirectMemory = metric.usedDirectMemory();
+        List<PoolArenaMetric> heapArenaMetrics = metric.heapArenas();
+        List<PoolArenaMetric> directArenaMetrics = metric.directArenas();
+        numHeapTotalAllocations = heapArenaMetrics.stream().mapToLong(PoolArenaMetric::numAllocations).sum();
+        numHeapTotalDeallocations = heapArenaMetrics.stream().mapToLong(PoolArenaMetric::numDeallocations).sum();
+        numHeapTotalActiveAllocations =
+            heapArenaMetrics.stream().mapToLong(PoolArenaMetric::numActiveAllocations).sum();
+        numDirectTotalAllocations = directArenaMetrics.stream().mapToLong(PoolArenaMetric::numAllocations).sum();
+        numDirectTotalDeallocations = directArenaMetrics.stream().mapToLong(PoolArenaMetric::numDeallocations).sum();
+        numDirectTotalActiveAllocations =
+            directArenaMetrics.stream().mapToLong(PoolArenaMetric::numActiveAllocations).sum();
+
+        try {
+          Thread.sleep(config.nettyMetricsRefreshIntervalSeconds * 1000);
```

Comment:
Can we use a `ScheduledExecutorService` to do this? Feel free to refer to `HistogramDumper` in `NonBlockingRouterMetrics`

Reply:
+1

Reply:
updated